docs/open-source/langchain-agent.mdx

```diff
@@ -1,209 +1,125 @@
 
-### Tool to get all contacts
-```python
-@tool("get_all_contacts")
-def get_all_contacts(placeholder: str) -> List[dict]:
-    """Get contacts."""
-    return CONTACTS
-```
+For example, we will design an agent that makes a custom chain querying Gemini Pro to make a poem based on the model name in the agent config:
 
-### Tool to call phone number
-`tools/vocode.py` makes use of the `OutboundCall` class to initiate a phone call
 ```python
-@tool("call phone number")
-def call_phone_number(input: str) -> str:
-    """calls a phone number as a bot and returns a transcript of the conversation.
-    the input to this tool is a pipe separated list of a phone number, a prompt, and the first thing the bot should say.
-    The prompt should instruct the bot with what to do on the call and be in the 3rd person,
-    like 'the assistant is performing this task' instead of 'perform this task'.
-
-    should only use this tool once it has found an adequate phone number to call.
-
-    for example, `+15555555555|the assistant is explaining the meaning of life|i'm going to tell you the meaning of life` will call +15555555555, say 'i'm going to tell you the meaning of life', and instruct the assistant to tell the human what the meaning of life is.
-    """
-    phone_number, prompt, initial_message = input.split("|", 2)
-    call = OutboundCall(
-        base_url=os.environ["TELEPHONY_SERVER_BASE_URL"],
-        to_phone=phone_number,
-        from_phone=os.environ["OUTBOUND_CALLER_NUMBER"],
-        config_manager=RedisConfigManager(),
-        agent_config=ChatGPTAgentConfig(
-            prompt_preamble=prompt,
-            initial_message=BaseMessage(text=initial_message),
-        ),
-        logger=logging.Logger("call_phone_number"),
-    )
-    LOOP.run_until_complete(call.start())
-    while True:
-        maybe_transcript = get_transcript(call.conversation_id)
-        if maybe_transcript:
-            delete_transcript(call.conversation_id)
-            return maybe_transcript
+from vocode.streaming.agent.abstract_factory import AbstractAgentFactory
+from vocode.streaming.models.agent import LangchainAgentConfig
+from vocode.streaming.agent.langchain_agent import LangchainAgent
+from langchain_google_genai import ChatGoogleGenerativeAI
+from langchain_core.prompts import ChatPromptTemplate
+
+class PoemLangchainAgent(LangchainAgent):
+    def create_chain(self):
+        prompt_template = ChatPromptTemplate.from_template(f"Make a random poem about {self.agent_config.model_name}")
+        model = ChatGoogleGenerativeAI(model="gemini-pro")
+        chain = prompt_template | model
+        return chain
+
+
+class MyAgentFactory(AbstractAgentFactory):
+    def create_agent(
+        self, agent_config: AgentConfig, logger: Optional[logging.Logger] = None
+    ) -> BaseAgent:
+        if isinstance(agent_config, LangchainAgentConfig):
```

Comment:
i wonder if this is the best way to dispatch into `PoemLangchainAgent` since `provider` is usually used to switch into the LLM integration on the langchain side. what if we created a separate `PoemLangchainAgentConfig` and dispatched into that via `isinstance`?

Reply:
Doesn't matter too much, I just tried to make an extremely bare-bones example but I can create a new config too.